File: 프로그래머스/3/388354. 홀짝트리/홀짝트리.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    
    private static final Map<Integer, List<Integer>> tree = new HashMap<>();
    
    private static boolean[] visited;
    
    public int[] solution(int[] nodes, int[][] edges) {   
        int[] answer = new int[2];
        
        for (int node : nodes) {
            tree.put(node, new ArrayList<>());
        }
        
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            tree.get(u).add(v);
            tree.get(v).add(u);
        }
        
        visited = new boolean[1000001];
        for (int root : nodes) {
            if (!visited[root] && dfs(root, -1, false)) answer[0]++;
        }
        
        visited = new boolean[1000001];
        for (int root : nodes) {
            if (!visited[root] && dfs(root, -1, true)) answer[1]++;
        }
               
        return answer;
    }
    
    private static boolean dfs(int curr, int prev, boolean isReversed) {
        List<Integer> children = tree.get(curr);
        int count = children.size();
        
        if (prev != -1) count--;
        if (isReversed && curr % 2 == count % 2) return false;
        if (!isReversed && curr % 2 != count % 2) return false;
        
        visited[curr] = true;
        
        for (int next : children) {
            if (next != prev && !dfs(next, curr, isReversed)) {
                visited[curr] = false;
                return false;
            }
        }
        
        return true;
    }
}

File: 프로그래머스/3/388354. 홀짝트리/홀짝트리.py
import sys

from collections import defaultdict

sys.setrecursionlimit(10**5)


def is_oe_tree(node, parent, visited, graph):
    children = graph[node]
    count = len(children)
    
    if parent != -1:
        count -= 1
    
    if node % 2 != count % 2:
        return False
    
    visited[node] = True
    
    for child in children:
        if child == parent:
            continue
        if not is_oe_tree(child, node, visited, graph):
            visited[node] = False
            return False
    
    return True


def is_roe_tree(node, parent, visited, graph):
    children = graph[node]
    count = len(children)
    
    if parent != -1:
        count -= 1
    
    if node % 2 == count % 2:
        return False
        
    visited[node] = True
    
    for child in children:
        if child == parent:
            continue
        if not is_roe_tree(child, node, visited, graph):
            visited[node] = False
            return False

    return True


def solution(nodes, edges):
    graph = defaultdict(list)
    for a, b in edges:
        graph[a].append(b)
        graph[b].append(a)
    
    visited = [False] * 1_000_001
    oe = sum(not visited[n] and is_oe_tree(n, -1, visited, graph) for n in nodes)
    
    visited = [False] * 1_000_001
    roe = sum(not visited[n] and is_roe_tree(n, -1, visited, graph) for n in nodes)
    
    return oe, roe